### PR TITLE
Create an application factory for flask #129

### LINKED ISSUE #129: Create an Application Factory for the Flask UI
Part of #128

Create an application factory for the flask ui and move the [module level setup into the factory](https://github.com/kizniche/Mycodo/blob/575b6548c0d1e716c8a4fb5d122be37b315393b1/mycodo/mycodo_flask/app.py#L120-L130)

[Basic Application Factory](http://flask.pocoo.org/docs/0.11/patterns/appfactories/)
```python
def create_app(config_filename):
    app = Flask(__name__)
    app.config.from_pyfile(config_filename)

    from yourapplication.model import db
    db.init_app(app)

    from yourapplication.views.admin import admin
    from yourapplication.views.frontend import frontend
    app.register_blueprint(admin)
    app.register_blueprint(frontend)

    return app
```

**WARNING**
The entire ```app.py``` module requres the app for setting up routes which are too delicate to deal with in this step.  These routes will eventually be rebuilt into small manageable blueprints which will make it much easier to understand what they are doing and keep isolated from each other but this is going to take time.  I suggest that the guts of the app.py module be thrown into a separate general module called ```routes``` with it's own blueprint.  We can attach it to the app in the app factory which will allow us the time needed to  slowly break up these routes.

### PR DESCRIPTION
**Pull request by etiology**
Fixes #129 

#### What does this update do?

This update reorganizes the flask UI app into an application factory pattern.  This pattern improves the application setup by separating it from the implementation and by breaking up the application configuration into functional groups.

Application setup is now performed via the ```create_app(config=ProdConfig)``` function [which returns a configured flask app](https://github.com/etiology/Mycodo/blob/9d81cfc3d8d1fd889b33c2447f8739e83fda0aea/mycodo/mycodo_flask/app.py#L42-L57).

Routes and Extensions are then registered to the app in separate functions:
 - [register_extensions](https://github.com/etiology/Mycodo/blob/9d81cfc3d8d1fd889b33c2447f8739e83fda0aea/mycodo/mycodo_flask/app.py#L60-L73)
 - [register_blueprints](https://github.com/etiology/Mycodo/blob/9d81cfc3d8d1fd889b33c2447f8739e83fda0aea/mycodo/mycodo_flask/app.py#L76-L79)

#### Other Modifications
 - [New tests](https://github.com/etiology/Mycodo/blob/9d81cfc3d8d1fd889b33c2447f8739e83fda0aea/mycodo/tests/software_tests/test_database/test_create_dbs.py#L1-L62) were added and the [fixtures were updated](https://github.com/etiology/Mycodo/blob/9d81cfc3d8d1fd889b33c2447f8739e83fda0aea/mycodo/tests/software_tests/test_mycodo_flask/conftest.py#L34-L48) to use the new application factory.  
 - Updated **Apache wsgi** file to [use the new app factory](https://github.com/etiology/Mycodo/blob/0e60103be1f5b6f079530d3ec38e37b94d2a9787/mycodo_flask.wsgi#L5).
 - [create_dbs](https://github.com/etiology/Mycodo/blob/0e60103be1f5b6f079530d3ec38e37b94d2a9787/init_databases.py#L126-L172) was modified to use the same object config that the flask app is using
 - Incremented the version number [1](https://github.com/etiology/Mycodo/blob/0e60103be1f5b6f079530d3ec38e37b94d2a9787/README.md#latest-version-4024), [2](https://github.com/etiology/Mycodo/blob/0e60103be1f5b6f079530d3ec38e37b94d2a9787/mycodo/config.py#L27)

For live testing I've installed this on a clean boot image and created a simple CPU sensor.  I was able to generate graphs and use the nav bar links.

### CODE REVIEW

**Comment by kizniche** [comment]
Looks great! The upgrade through the UI was smooth and without a single error.

**Comment by kizniche** [comment]
I just updated one of my Pis and my HTTP log started becoming populated with some errors (the server starts and runs).

I started the update at 18:01:37, then the following appeared in the log. I then restarted apache with ```sudo /etc/init.d/apache2 restart``` at 19:50:59.

```
/var/log/apache2/error.log:

[Wed Oct 26 18:04:02.495019 2016] [wsgi:error] [pid 26980:tid 3036648496] 32.5678436673
[Wed Oct 26 18:04:03.679585 2016] [wsgi:error] [pid 26980:tid 3036648496] Creating/verifying None at sqlite:////home/kiz/Mycodo/mycodo/../databases/mycodo.db ...
[Wed Oct 26 18:04:10.693739 2016] [wsgi:error] [pid 26980:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: alembic_version.version_num [SQL: u'INSERT INTO alembic_version (version_num) VALUES (?)'] [parameters: ('52aeaa701dfb',)]
[Wed Oct 26 18:04:11.137583 2016] [wsgi:error] [pid 26980:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: displayorder.id [SQL: u'INSERT INTO displayorder (id, graph, lcd, log, pid, relay, remote_host, sensor, timer) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', '', None, '', '', '', None, '', None)]
[Wed Oct 26 18:04:11.589038 2016] [wsgi:error] [pid 26980:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: cameratimelapse.id [SQL: u'INSERT INTO cameratimelapse (id, relay_id, path, prefix, file_timestamp, display_last, cmd_pre_camera, cmd_post_camera, extra_parameters) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', '', '/var/www/mycodo/camera-timelapse', 'Timelapse', 1, 1, '', '', '--nopreview --contrast 20 --sharpness 60 --awb auto --quality 20 --vflip --hflip --width 800 --height 600')]
[Wed Oct 26 18:04:11.909287 2016] [wsgi:error] [pid 26980:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: camerastill.id [SQL: u'INSERT INTO camerastill (id, hflip, vflip, rotation, relay_id, timestamp, display_last, cmd_pre_camera, cmd_post_camera, extra_parameters) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', 0, 0, 0, '', 1, 1, '', '', '--vflip --hflip --width 800 --height 600')]
[Wed Oct 26 18:04:12.300504 2016] [wsgi:error] [pid 26980:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: camerastream.id [SQL: u'INSERT INTO camerastream (id, relay_id, cmd_pre_camera, cmd_post_camera, extra_parameters) VALUES (?, ?, ?, ?, ?)'] [parameters: ('0', '', '', '', '--contrast 20 --sharpness 60 --awb auto --quality 20 --vflip --hflip --nopreview --width 800 --height 600')]
[Wed Oct 26 18:04:12.678036 2016] [wsgi:error] [pid 26980:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: misc.id [SQL: u'INSERT INTO misc (id, force_https, dismiss_notification, hide_alert_success, hide_alert_info, hide_alert_warning, stats_opt_out, login_message, relay_stats_volts, relay_stats_cost, relay_stats_currency, relay_stats_dayofmonth) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', 1, 0, 0, 0, 0, 0, '', 120, 0.05, '$', 15)]
[Wed Oct 26 18:04:12.931264 2016] [wsgi:error] [pid 26980:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: smtp.id [SQL: u'INSERT INTO smtp (id, host, ssl, port, user, passw, email_from, hourly_max) VALUES (?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', 'smtp.gmail.com', 1, 465, 'email@gmail.com', 'password', 'email@gmail.com', 2)]
[Wed Oct 26 18:04:12.965363 2016] [wsgi:error] [pid 26980:tid 3036648496] Creating/verifying None at sqlite:////home/kiz/Mycodo/mycodo/../databases/notes.db ...
[Wed Oct 26 18:04:13.771296 2016] [wsgi:error] [pid 26980:tid 3036648496] Creating/verifying None at sqlite:////home/kiz/Mycodo/mycodo/../databases/users.db ...
[Wed Oct 26 18:11:27.847921 2016] [wsgi:error] [pid 27614:tid 3036648496] 33.3759813854
[Wed Oct 26 18:11:28.594281 2016] [wsgi:error] [pid 27614:tid 3036648496] Creating/verifying None at sqlite:////home/kiz/Mycodo/mycodo/../databases/mycodo.db ...
[Wed Oct 26 18:11:33.528209 2016] [wsgi:error] [pid 27614:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: alembic_version.version_num [SQL: u'INSERT INTO alembic_version (version_num) VALUES (?)'] [parameters: ('52aeaa701dfb',)]
[Wed Oct 26 18:11:33.664950 2016] [wsgi:error] [pid 27614:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: displayorder.id [SQL: u'INSERT INTO displayorder (id, graph, lcd, log, pid, relay, remote_host, sensor, timer) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', '', None, '', '', '', None, '', None)]
[Wed Oct 26 18:11:33.796268 2016] [wsgi:error] [pid 27614:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: cameratimelapse.id [SQL: u'INSERT INTO cameratimelapse (id, relay_id, path, prefix, file_timestamp, display_last, cmd_pre_camera, cmd_post_camera, extra_parameters) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', '', '/var/www/mycodo/camera-timelapse', 'Timelapse', 1, 1, '', '', '--nopreview --contrast 20 --sharpness 60 --awb auto --quality 20 --vflip --hflip --width 800 --height 600')]
[Wed Oct 26 18:11:33.912718 2016] [wsgi:error] [pid 27614:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: camerastill.id [SQL: u'INSERT INTO camerastill (id, hflip, vflip, rotation, relay_id, timestamp, display_last, cmd_pre_camera, cmd_post_camera, extra_parameters) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', 0, 0, 0, '', 1, 1, '', '', '--vflip --hflip --width 800 --height 600')]
[Wed Oct 26 18:11:34.046677 2016] [wsgi:error] [pid 27614:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: camerastream.id [SQL: u'INSERT INTO camerastream (id, relay_id, cmd_pre_camera, cmd_post_camera, extra_parameters) VALUES (?, ?, ?, ?, ?)'] [parameters: ('0', '', '', '', '--contrast 20 --sharpness 60 --awb auto --quality 20 --vflip --hflip --nopreview --width 800 --height 600')]
[Wed Oct 26 18:11:34.252997 2016] [wsgi:error] [pid 27614:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: misc.id [SQL: u'INSERT INTO misc (id, force_https, dismiss_notification, hide_alert_success, hide_alert_info, hide_alert_warning, stats_opt_out, login_message, relay_stats_volts, relay_stats_cost, relay_stats_currency, relay_stats_dayofmonth) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', 1, 0, 0, 0, 0, 0, '', 120, 0.05, '$', 15)]
[Wed Oct 26 18:11:34.426244 2016] [wsgi:error] [pid 27614:tid 3036648496] (sqlite3.IntegrityError) UNIQUE constraint failed: smtp.id [SQL: u'INSERT INTO smtp (id, host, ssl, port, user, passw, email_from, hourly_max) VALUES (?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', 'smtp.gmail.com', 1, 465, 'email@gmail.com', 'password', 'email@gmail.com', 2)]
[Wed Oct 26 18:11:34.437497 2016] [wsgi:error] [pid 27614:tid 3036648496] Creating/verifying None at sqlite:////home/kiz/Mycodo/mycodo/../databases/notes.db ...
[Wed Oct 26 18:11:35.011031 2016] [wsgi:error] [pid 27614:tid 3036648496] Creating/verifying None at sqlite:////home/kiz/Mycodo/mycodo/../databases/users.db ...
[Wed Oct 26 19:50:59.424135 2016] [mpm_event:notice] [pid 790:tid 3069383184] AH00491: caught SIGTERM, shutting down
[Wed Oct 26 19:51:04.688355 2016] [ssl:warn] [pid 2435:tid 3069305360] AH01906: 127.0.1.1:443:0 server certificate is a CA certificate (BasicConstraints: CA == TRUE !?)
[Wed Oct 26 19:51:04.690461 2016] [ssl:warn] [pid 2435:tid 3069305360] AH01909: 127.0.1.1:443:0 server certificate does NOT include an ID which matches the server name
[Wed Oct 26 19:51:05.250748 2016] [ssl:warn] [pid 2436:tid 3069305360] AH01906: 127.0.1.1:443:0 server certificate is a CA certificate (BasicConstraints: CA == TRUE !?)
[Wed Oct 26 19:51:05.251132 2016] [ssl:warn] [pid 2436:tid 3069305360] AH01909: 127.0.1.1:443:0 server certificate does NOT include an ID which matches the server name
[Wed Oct 26 19:51:05.255519 2016] [wsgi:warn] [pid 2436:tid 3069305360] mod_wsgi: Compiled for Python/2.7.8.
[Wed Oct 26 19:51:05.255714 2016] [wsgi:warn] [pid 2436:tid 3069305360] mod_wsgi: Runtime using Python/2.7.9.
[Wed Oct 26 19:51:05.274987 2016] [mpm_event:notice] [pid 2436:tid 3069305360] AH00489: Apache/2.4.10 (Raspbian) OpenSSL/1.0.1t mod_wsgi/4.3.0 Python/2.7.9 configured -- resuming normal operations
[Wed Oct 26 19:51:05.275304 2016] [core:notice] [pid 2436:tid 3069305360] AH00094: Command line: '/usr/sbin/apache2'
[Wed Oct 26 19:51:49.267047 2016] [wsgi:error] [pid 2439:tid 3028182064] 36.8430463734
[Wed Oct 26 19:51:49.797197 2016] [wsgi:error] [pid 2439:tid 3028182064] Creating/verifying None at sqlite:////home/kiz/Mycodo/mycodo/../databases/mycodo.db ...
[Wed Oct 26 19:51:56.749100 2016] [wsgi:error] [pid 2439:tid 3028182064] (sqlite3.IntegrityError) UNIQUE constraint failed: alembic_version.version_num [SQL: u'INSERT INTO alembic_version (version_num) VALUES (?)'] [parameters: ('52aeaa701dfb',)]
[Wed Oct 26 19:51:56.981077 2016] [wsgi:error] [pid 2439:tid 3028182064] (sqlite3.IntegrityError) UNIQUE constraint failed: displayorder.id [SQL: u'INSERT INTO displayorder (id, graph, lcd, log, pid, relay, remote_host, sensor, timer) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', '', None, '', '', '', None, '', None)]
[Wed Oct 26 19:51:57.218981 2016] [wsgi:error] [pid 2439:tid 3028182064] (sqlite3.IntegrityError) UNIQUE constraint failed: cameratimelapse.id [SQL: u'INSERT INTO cameratimelapse (id, relay_id, path, prefix, file_timestamp, display_last, cmd_pre_camera, cmd_post_camera, extra_parameters) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', '', '/var/www/mycodo/camera-timelapse', 'Timelapse', 1, 1, '', '', '--nopreview --contrast 20 --sharpness 60 --awb auto --quality 20 --vflip --hflip --width 800 --height 600')]
[Wed Oct 26 19:51:57.474515 2016] [wsgi:error] [pid 2439:tid 3028182064] (sqlite3.IntegrityError) UNIQUE constraint failed: camerastill.id [SQL: u'INSERT INTO camerastill (id, hflip, vflip, rotation, relay_id, timestamp, display_last, cmd_pre_camera, cmd_post_camera, extra_parameters) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', 0, 0, 0, '', 1, 1, '', '', '--vflip --hflip --width 800 --height 600')]
[Wed Oct 26 19:51:57.830487 2016] [wsgi:error] [pid 2439:tid 3028182064] (sqlite3.IntegrityError) UNIQUE constraint failed: camerastream.id [SQL: u'INSERT INTO camerastream (id, relay_id, cmd_pre_camera, cmd_post_camera, extra_parameters) VALUES (?, ?, ?, ?, ?)'] [parameters: ('0', '', '', '', '--contrast 20 --sharpness 60 --awb auto --quality 20 --vflip --hflip --nopreview --width 800 --height 600')]
[Wed Oct 26 19:51:58.100827 2016] [wsgi:error] [pid 2439:tid 3028182064] (sqlite3.IntegrityError) UNIQUE constraint failed: misc.id [SQL: u'INSERT INTO misc (id, force_https, dismiss_notification, hide_alert_success, hide_alert_info, hide_alert_warning, stats_opt_out, login_message, relay_stats_volts, relay_stats_cost, relay_stats_currency, relay_stats_dayofmonth) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', 1, 0, 0, 0, 0, 0, '', 120, 0.05, '$', 15)]
[Wed Oct 26 19:51:58.330418 2016] [wsgi:error] [pid 2439:tid 3028182064] (sqlite3.IntegrityError) UNIQUE constraint failed: smtp.id [SQL: u'INSERT INTO smtp (id, host, ssl, port, user, passw, email_from, hourly_max) VALUES (?, ?, ?, ?, ?, ?, ?, ?)'] [parameters: ('0', 'smtp.gmail.com', 1, 465, 'email@gmail.com', 'password', 'email@gmail.com', 2)]
[Wed Oct 26 19:51:58.350382 2016] [wsgi:error] [pid 2439:tid 3028182064] Creating/verifying None at sqlite:////home/kiz/Mycodo/mycodo/../databases/notes.db ...
[Wed Oct 26 19:51:58.990547 2016] [wsgi:error] [pid 2439:tid 3028182064] Creating/verifying None at sqlite:////home/kiz/Mycodo/mycodo/../databases/users.db ...
```

**Comment by etiology** [comment]
Ok so I'm still looking into this but this is what I see:
  - ```IntegrityError```'s on the unique constraints is likely because the database already exists which would occur anytime this is not the first call to create_dbs.
  - Lines that contain ```Creating/verifying None at sqlite:/...``` are coming from the [print statements in this area](https://github.com/kizniche/Mycodo/blob/b8123afbc2840751626b474c48984884f141220a/init_databases.py#L151-L166)
  - ```certificate does NOT include an ID which matches the server name``` I'm not sure about yet but I'll look into it.
 
[create_dbs](https://github.com/kizniche/Mycodo/blob/b8123afbc2840751626b474c48984884f141220a/init_databases.py#L126) is getting called when the WSGI [app](https://github.com/kizniche/Mycodo/blob/b8123afbc2840751626b474c48984884f141220a/mycodo/mycodo_flask/app.py#L65) is getting setup.  It shouldn't affect anything [because it doesn't replace the databases or the data in them](http://docs.sqlalchemy.org/en/latest/core/metadata.html#sqlalchemy.schema.MetaData.create_all).  It does mean that the UI now has the ability to build the databases if they are missing instead of crashing and requiring the user to run the init_databases again.

It's a bit wordy at the moment due to the print statements and perhaps we should replace them with debug level logging message.  

What are you thoughts/feelings on this?

**Comment by kizniche** [comment]
Ah, [that's](https://github.com/kizniche/Mycodo/blob/b8123afbc2840751626b474c48984884f141220a/mycodo/mycodo_flask/app.py#L65) where create_dbs is being called from.

I don't really mind it appearing in the apache log, but maybe some checks should be included to determine if the dbs need to be created instead of just trying, like in [test_create_dbs.py](https://github.com/kizniche/Mycodo/blob/master/mycodo/tests/software_tests/test_database/test_create_dbs.py#L41-L42).

**Comment by etiology** [comment]
We can perform those checks but keep in mind that those checks are performed by sqlalchemy with a much better precision. We could delete a single table out of one of the three databases and sqlalchemy will only rebuild the missing table.  

SQLAlchemy's pretty surgical about it. Right now it talks too much due to the prints, but it's similar to what flask-sqlalchemy behavior is which is replace any missing piece when the UI loads.

**Comment by kizniche** [comment]
I'm trying to think about a situation where the database would exist yet not have the proper tables, and if I would want a more vocal indication of that (server error) instead of silently fixing the issue.

**Comment by etiology** [comment]
We can remove it if needed.  Tests would need a small modification to create the databases before the app is fully created.  The Flask UI does a [check on the user database](https://github.com/kizniche/Mycodo/blob/b8123afbc2840751626b474c48984884f141220a/mycodo/mycodo_flask/general_routes.py#L102-L115) and fails if it doesn't exist.

I'm just thinking out loud here but what about a future option of having the admin user setup moved to the UI where it's nice and pretty?  The user could just run the setup.py code, get a receipt number and use it in a pretty form that appears when the system isn't setup yet.  They could enter their credentials and add the receipt number to verify that they have shell access.  Kind of reminds me of setting up a new computer.  Just a thought.  

Let me know if you want the ```create_dbs```behavior moved out of the Flask UI factory and I'll make sure the tests are updated too.

**Comment by kizniche** [comment]
I don't think it should be removed but agree the prints should be cleaned up or at least all expected messages should maybe be logged at debug level.

I like the idea of creating the admin user through the UI. Would detecting the client IP of 127.0.0.1 be sufficient for credentials to create an admin?

**Comment by etiology** [comment]
That's a clever idea.  It would require that you have a monitor setup to your raspberry pi and you'd be using a local browser but I could totally see that as the default *of course you are authorized* behavior.  Could really be a nice finishing touch.  I think most people use a monitor.  I don't have an extra one so I only use shell access to set up the pi so I would need the alternative route as a backup.

The receipt reminds me of how the debug works now in flask where you can enable the console via the error page if you can type the correct code.

**Comment by kizniche** [comment]
Ah, you're right. It would more likely be an IP of your router that would be connecting. It could include all of the private address ranges (192.168.0, 172.16.0, 10.0.0,127.0.0.1). If connecting from anything else, it could ask for the receipt.

**Comment by kizniche** [comment]
I'm wondering if we really even need to worry about this security. The person installing Mycodo will most likely go to the web address after running the setup.

Is there a way to find the receipt later if the user didn't save it?

**Comment by etiology** [comment]
I think I was over thinking it. In the past I've used a manage.py script to run admin commands, but then I remembered that [Flask added a command shell](http://flask.pocoo.org/docs/0.11/cli/). 

That can be used to interact with and set up the database. Simple fix.

**Comment by kizniche** [comment]
But we still need a way to add at least one admin username/password. This is currently done by manually executing the database init script. We could just have this database created and an admin user created as the first part of the setup script.

**Comment by etiology** [comment]
At flask command:
```python
new_admin = User(name="Joe", email="joe@email.com", admin=True)

new_admin.set_password("super_safe")
new_admin.save()
```

**Comment by kizniche** [comment]
And this will be populated by a form prompted to the user?

**Comment by etiology** [comment]
Sorry I should have added more context.  I was on my phone so it was all thumbs.  The new_admin creation is an example that you would see when you use a tool like [flask-script](https://flask-script.readthedocs.io/en/latest/).  When you SSH into your server (raspberry pi) you run the script and can pass it commands.  For instance if you set this up with a customized shell environment that is populated with your database models (like the User sqlalchemy model):
```python manager.py shell```

Since the shell already knows about my ```User``` model I can just create a new user like I did in the example above.  The user model has a boolean attribute to denote that this user is an admin.  This way you can have multiple admins.  Using my new user object ```new_admin``` I can setup a password and save the user calling a CRUD method ```save()```.

So with only shell access I can setup a user or interact with the model data of my database.  We could also just keep the script that is already here, but again I was just thinking out loud with the admin user being configured via the web interface. 

The app currently uses a shell script and we can just add a UI form and admin setup page.  I could imagine that we modify the behavior in the [before_request](https://github.com/kizniche/Mycodo/blob/b8123afbc2840751626b474c48984884f141220a/mycodo/mycodo_flask/general_routes.py#L102-L115) method and redirect them to the admin creation page.  

Later if this project uses [Flask-Login](https://flask-login.readthedocs.io/en/latest/) or the larger [Flask-Security](https://pythonhosted.org/Flask-Security/) we can have these functions redirect to the same admin creation page.

**Comment by kizniche** [comment]
I just pushed commit dbb53fb for the [admin_creation](https://github.com/kizniche/Mycodo/tree/admin_creation) branch. I've tested it and can't find any security flaws. Take a look and let me know if this is what you were thinking.

It will display a form to create an admin if there are no admin users in the user database. Then redirect to the login form if an admin user was successfully created.

**Comment by kizniche** [comment]
I don't believe I implemented your user addition method, though. I wasn't able to fully figure it out, so I just copied the previous method from the UI user creation code.

**Comment by kizniche** [comment]
Also, this was the first commit from PyCharm, using the remote deployment feature. It was suspiciously easy to work with.

**Comment by etiology** [comment]
> I don't believe I implemented your user addition method, though. I wasn't able to fully figure it out, so I just copied the previous method from the UI user creation code.

The user model would need to change before that.  At some point I'd like to review the database tables with you and get an idea of how everything should and is fitting together.  I'd like to move towards pushing some of the integrity checks and consistency constraints to the database.

I'm excited to take a look at the new admin creation page.  Not sure if I can get to it tonight but it sounds like good.
